draco/analysis/ringmapmaker.py

```diff
@@ -411,25 +419,149 @@ class RingMapMaker(task.group_tasks(MakeVisGrid, BeamformNS, BeamformEW)):
+
+        rm.weight[:] = 0.0
+        rm.redistribute("freq")
+
+        # Dereference datasets
+        hv = hybrid_vis_m.vis[:].view(np.ndarray)
+        hw = hybrid_vis_m.weight[:].view(np.ndarray)
+
+        rmm = rm.map[:]
+        rmw = rm.weight[:]
+        if self.save_dirty_beam:
+            rmb = rm.dirty_beam[:]
+
+        # Loop over frequencies
+        for lfi, freq in enumerate(local_freq):
+
+            find = (slice(None),) * 3 + (lfi,)
+
+            hvf = hv[find]
+
+            # Make copy here because we will modify weights if exclude_intracyl is set
+            inv_var = hw[find][..., np.newaxis].copy()
+
+            # Get the EW weights using method defined by subclass
+            weight = self._get_weight(inv_var) * (inv_var > 0.0)
+
+            # Get the beam m-modes using method defined by subclass
+            bvf = self._get_beam_mmodes(freq, hybrid_vis_m)
+
+            # Get the regularisation term, exact prescription is defined by the subclass
+            epsilon = self._get_regularisation(freq)
+
+            # Calculate the normalization
+            sum_weight = (weight * np.abs(bvf) ** 2).sum(axis=(1, -2))
+
+            C_inv = epsilon + sum_weight
+
+            # Solve for the sky m-modes
+            map_m = (bvf.conj() * weight * hvf).sum(
+                axis=(1, -2)
+            ) * tools.invert_no_zero(C_inv)
+
+            # Calculate the dirty beam m-modes
+            dirty_beam_m = sum_weight * tools.invert_no_zero(C_inv)
+
+            # Calculate the point source normalization (dirty beam at transit)
+            norm = tools.invert_no_zero(dirty_beam_m.mean(axis=0))[:, np.newaxis, :]
+
+            # Fill in the ringmap
+            rmm[0, :, lfi] = (
+                np.fft.irfft(map_m.transpose(1, 2, 0), axis=-1, n=nra).transpose(
+                    0, 2, 1
+                )
+                * norm
+            )
+
+            # Fill in the dirty beam
+            if self.save_dirty_beam:
+                rmb[0, :, lfi] = (
+                    np.fft.irfft(
+                        dirty_beam_m.transpose(1, 2, 0), axis=-1, n=nra
+                    ).transpose(0, 2, 1)
+                    * norm
+                )
+
+            # Calculate the expected map noise by propagating the uncertainty on the m's
+            var = tools.invert_no_zero(inv_var)
+            var = ((weight * np.abs(bvf)) ** 2 * var).sum(
+                axis=(1, -2)
+            ) * tools.invert_no_zero(C_inv ** 2)
+            sum_var_map_m = 0.5 * np.sum(var, axis=0)[:, np.newaxis, :]
+
+            rmw[:, lfi] = (mmax + 1) ** 2 * tools.invert_no_zero(
+                norm ** 2 * sum_var_map_m
+            )
+
+        return rm
+
+    def _get_beam_mmodes(self, freq, hybrid_vis_m):
+        raise NotImplementedError
+
+    def _get_weight(self, inv_var):
+        raise NotImplementedError
+
+    def _get_regularisation(self, freq):
+        raise NotImplementedError
```

Comment:
It would be good to add docstrings in here so people know what they would need to implement to make a full deconvolved ringmap class.